sealir-tutorials/ch06_2_typeinfer_loops.py

```diff
@@ -0,0 +1,268 @@
+        union(TypeVar(operands[idx])).with_(TypedIns(region).arg(idx)),
+    )
+
+    yield rule(
+        loop == Term.Loop(body=body, operands=TermList(operands)),
+        body == Term.RegionEnd(region=region, ports=ports),
+    ).then(
+        # propagate loop outputs
+        assign_output_loop_typevar(0, operands.length(), ports, operands, loop)
+    )
+
+    yield rule(
+        assign_output_loop_typevar(start, stop, ports, operands, loop),
+        start + 1 < stop,
+    ).then(
+        assign_output_loop_typevar(start + 1, stop, ports, operands, loop),
+    )
+
+    yield rule(
+        assign_output_loop_typevar(start, stop, ports, operands, loop),
+        start > 0,
+    ).then(
+        # TypeVars of loop-region output ports are loop output ports
+        union(TypeVar(ports.getValue(start))).with_(
+            # minus 1 in because loop output drop the loop condition
+            TypeVar(loop.getPort(start - 1))
+        ),
+        # TypeVars of loop region output ports are the same the operands
+        union(TypeVar(ports.getValue(start))).with_(
+            # minus 1 in because loop inputs do not have the loop condition
+            TypeVar(operands[start - 1])
+        ),
+    )
+
+
+# ### Define rulesets for extra operations needed:
+
+
+@ruleset
+def ruleset_type_infer_undef(x: Term, y: Term, io: Term):
+    yield rule(
+        # Undef operations that are typed to Int64 becomes
+        # a literal i64 0
+        x == Term.Undef(_wc(String)),
+        TypeVar(x).getType() == TypeInt64,  # output is Int64
+    ).then(union(x).with_(Term.LiteralI64(0)))
+    yield rule(
+        # Undef operations that are typed to Bool becomes
+        # a literal bool 0
+        x == Term.Undef(_wc(String)),
+        TypeVar(x).getType() == TypeBool,  # output is Bool
+    ).then(union(x).with_(Term.LiteralBool(False)))
+
+
+@function
+def Nb_Not_Int64(operand: Term) -> Term: ...
+
+
+@ruleset
+def ruleset_type_infer_not(x: Term, y: Term, io: Term):
+    yield rule(
+        # Type-infer unary not that takes a Int64
+        y == Py_NotIO(io=io, term=x),
+        TypeVar(x).getType() == TypeInt64,
+    ).then(
+        # Shortcut IO
+        union(y.getPort(0)).with_(io),
+        # The result becomes Nb_Not_Int64
+        union(y.getPort(1)).with_(Nb_Not_Int64(x)),
+        # Output is Bool
+        set_(TypeVar(Nb_Not_Int64(x)).getType()).to(TypeBool),
+    )
+
+
+# ### Extend EGraphToRVSDG class from ch6.1 to handle the extra operations
+
+
+class NbOp_Not_Int64(NbOp_Base):
+    operand: SExpr
+
+
+class ExtendEGraphToRVSDG(_ch06_1_ExtendEGraphToRVSDG):
+    def handle_Term(self, op: str, children: dict | list, grm: Grammar):
+        match op, children:
+            case "Nb_Not_Int64", {"operand": operand}:
+                return grm.write(NbOp_Not_Int64(operand=operand))
+        return super().handle_Term(op, children, grm)
+
+
+# ### Extend the LLVM Backend from ch6.1
+
+
+class Backend(ch06_1_Backend):
+    pass
```

Comment:
The loop logic is already implemented in 6.1. I don't think you need to split ch6 like I did for ch4. You can just have a single file for ch6